core/deployment/src/main/java/io/quarkus/deployment/dev/IsolatedDevModeMain.java

```diff
@@ -57,7 +60,31 @@ private synchronized void firstStart() {
             //ok, we have resolved all the deps
             try {
                 StartupAction start = augmentAction.createInitialRuntimeApplication();
-                runner = start.run();
+                //this is a bit yuck, but we need replace the default
+                //exit handler in the runtime class loader
+                //TODO: look at implementing a common core classloader, that removes the need for this sort of crappy hack
+                curatedApplication.getBaseRuntimeClassLoader().loadClass(ApplicationLifecycleManager.class.getName())
```

Comment:
Shouldn't this be done only in command mode? Or this class is only used for command mode?

Reply:
This is installed every time. For a normal hot reload the 'restarting' flag will be true, so nothing will happen.

If on the other hand something has called asyncExit or it is a command mode app and main has returned you get the option to press enter to re-run.